Sources/TimeFreeze.swift

```diff
@@ -70,18 +70,36 @@ struct TimeFreeze {
     /// see: https://github.com/darwin-on-arm/xnu/blob/master/osfmk/kern/clock.c#L522.
     ///
     /// - returns: An Int measurement of system uptime in microseconds.
+//    static func systemUptime() -> TimeInterval {
+//        var mib = [CTL_KERN, KERN_BOOTTIME]
+//        var size = MemoryLayout<timeval>.stride
+//        var bootTime = timeval()
+//
+//        let bootTimeError = sysctl(&mib, u_int(mib.count), &bootTime, &size, nil, 0) != 0
+//        assert(!bootTimeError, "system clock error: kernel boot time unavailable")
+//
+//        let now = currentTime()
+//        let uptime: TimeInterval = Double(bootTime.tv_sec) + Double(bootTime.tv_usec) / 1_000_000
+//        assert(now >= uptime, "inconsistent clock state: system time precedes boot time")
+//        
+//        return now - uptime
+//    }
+    
+    /// Returns a high-resolution measurement of system uptime, that continues ticking through device sleep
+    /// *and* user- or system-generated clock adjustments. This allows for stable differences to be calculated
+    /// between timestamps.
+    ///
+    /// - returns: An Int measurement of system uptime in nanoseconds.
+    /// - reference: https://stackoverflow.com/a/45068046/23930754
+    /// - Note: This function is thread-safe. Can be called from any thread concurrently.
     static func systemUptime() -> TimeInterval {
-        var mib = [CTL_KERN, KERN_BOOTTIME]
-        var size = MemoryLayout<timeval>.stride
-        var bootTime = timeval()
-
-        let bootTimeError = sysctl(&mib, u_int(mib.count), &bootTime, &size, nil, 0) != 0
-        assert(!bootTimeError, "system clock error: kernel boot time unavailable")
-
-        let now = Date().timeIntervalSince1970
-        let uptime = Double(bootTime.tv_sec) + Double(bootTime.tv_usec) / 1_000_000
-        assert(now >= uptime, "inconsistent clock state: system time precedes boot time")
-
-        return now - uptime
+        var uptime = timespec()
+        /// Thread-safe reference: https://man7.org/linux/man-pages/man3/clock_gettime.3.html#ATTRIBUTES
```

Comment:
The link is Linux manual page, which does not strictly apply to Unix systems in general.